### PR TITLE
CA-224730 back off exponentially after failed read from plugin

### PR DESCRIPTION
**Pull request by lindig**
This commit introduces a skip count for plugins. After a read error, the
skip count is incremented and the plugin is not read for that number of
read cycles. When reading succeeds, the skip is reset to 0. Currently,
at most 256 cycles are skipped and the skip count increases
exponentially from 1 with every read error.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

### CODE REVIEW

**Comment by euanh** [comment]
In the current implementation, the functions for handling the skip counts are mixed in with the functions for reading and writing the RRDs.   I think it would be cleaner to move the skip mechanism out into a module of its own.

The logic for deciding when to skip also intrudes into the RRD handling code.   Have you tried making the skip function into a wrapper or decorator which could be applied to the RRD-handling function?  It could catch the exception generated by the underlying function, or look at a return code, and decide whether to log, stay silent, or reset the counter.

**Comment by mseri** [comment]
Looks good! I had just a minor comment.

File: rrdd/rrdd_server.ml [hunk 1]
```diff
@@ -415,36 +416,44 @@ module Plugin = struct
...
+			end
+
+		(* we are skipping a reading *)
+		let decr_skip_count (uid, plugin) =
+			if plugin.skip > 0 then
+				Mutex.execute registered_m  (fun () ->
+					plugin.skip <- plugin.skip - 1)
+
+		(* true, iff the plugin skips the next reading *)
+		let skip (uid, plugin) = plugin.skip > 0
```
Comment (mseri) [inline rrdd/rrdd_server.ml]: Is this used anywhere?

**Comment by lindig** [comment]
The code lives inside a relatively small functor which constraints what the module can do. I am reluctant to introduce more modules because xcp-rrdd is already over-functorized.

File: rrdd/rrdd_server.ml (hunk 1 above)
Comment (lindig) [inline rrdd/rrdd_server.ml]: Yes - the at the call side further down. 

https://github.com/xapi-project/xcp-rrdd/pull/74/files#diff-60295083deed8eb9d4094ac3270fbec5R529

Comment (mseri) [inline rrdd/rrdd_server.ml]: Ah right, I missed it. I think you could define it earlier and use it in `decr_skip_count` (line 442) as well

Comment (lindig) [inline rrdd/rrdd_server.ml]: I've updated this.

**Comment by lindig** [comment]
I agree with @euanh that the functor that handles registering, unregistering and reading plugins could be better organised to accommodate back-off from plugins that fail. However, this amounts to re-writing this functor.

File: rrdd/rrdd_server.ml [hunk 2]
```diff
@@ -415,36 +416,44 @@ module Plugin = struct
...
-				warn "unmuting plugin %s" (P.string_of_uid uid)
-			end else
-				()
-
-		let get_payload ~(uid: P.uid) : Rrd_protocol.payload =
-			let plugin  = find uid in
+		(* we hit an error - increase skip count exponentially until max *)
+		let incr_skip_count uid plugin =
+			let skip_max  = 256 in (* about 21.3 min on 5sec cycle *)
+			let min x y   = if x < y then x else y in
```
Comment (robhoes) [inline rrdd/rrdd_server.ml]: The `min` function is defined in pervasives.

Comment (lindig) [inline rrdd/rrdd_server.ml]: Fixed it.

**Comment by lindig** [comment]
I have made a comment that the `Make` functor should be re-written when the next change comes along. I'd rather not do it on a CA ticket but I'm happy to come back to it even short term.

**Pull request merged**